src/noise.hpp

```diff
@@ -1,459 +1,504 @@
-					int xm = ((x-1) < 0) ? (N-2) : (x-1);
-					int yp = ((y+1) == N) ? 1 : (y+1);
-					int ym = ((y-1) < 0) ? (N-2) : (y-1);
-					neigh 	<< in(xm,ym) << in(x,ym) << in(xp,ym) << arma::endr
-							<< in(xm,y)  << in(x,y)  << in(xp,y) << arma::endr
-							<< in(xm,yp) << in(x,yp) << in(xp,yp) << arma::endr;
-
-					out(y,x) = arma::accu(neigh % (-1*laplacian));
-				}
-			}
-			return out;
-		};
-
-		// Recursive quadtree function
-		std::vector<arma::umat> QuadTree(const arma::mat &A, std::vector<arma::umat> treeDelete, int part) {
-
-			int i = treeDelete[0](0, part);
-			int j = treeDelete[0](1, part);
-			int s = treeDelete[0](2, part);
-
-			// Test if block should be split
-			arma::mat patch = A.submat(arma::span(i, i+s-1), arma::span(j, j+s-1));
-			if(!SplitBlockQ(patch)) {
-				return treeDelete;
-			}
-
-			// If test returns TRUE, split
-			s /= 2;
-			int n = treeDelete[0].n_cols - 1;
-
-			arma::umat newtreeadd(3,4);
-			newtreeadd	<< i << i+s << i << i+s << arma::endr
-						<< j << j << j+s << j+s << arma::endr
-						<< s << s << s << s << arma::endr;
-			arma::umat newtree = arma::join_horiz(treeDelete[0], newtreeadd);
-
-			arma::umat newdeleteadd(1,1);
-			newdeleteadd << part << arma::endr;
-			arma::umat newdelete = arma::join_horiz(treeDelete[1], newdeleteadd);
-
-			std::vector<arma::umat> newtreeDelete;
-			newtreeDelete.resize(2);
-			newtreeDelete[0] = newtree;
-			newtreeDelete[1] = newdelete;
-
-			int iter = n;
-			do {
-				newtreeDelete = QuadTree(A, newtreeDelete, iter);
-				iter++;
-			} while(iter < n+4);
-
-			return newtreeDelete;
-		};
-
+ public:
+        NoiseEstimator() {
+            // Set Laplacian kernel to 3x3
+            laplacian = 1/8 * arma::ones<arma::mat>(3,3);
+            laplacian(1,1) = -1;
+            treeDelete.resize(2);
+        };
+
+        ~NoiseEstimator() {};
+
+        void Estimate(const arma::cube &input,
+                            double &alphaIn,
+                            double &muIn,
+                            double &sigmaIn,
+                            int sizeIn) {
+            // Read from parameters
+            alpha = alphaIn;
+            mu = muIn;
+            sigma = sigmaIn;
+            size = sizeIn;
+
+            // Set some parameters
+            dSi = 0.;
+            Nx = input.n_cols;
+            Ny = input.n_rows;
+            T = input.n_slices;
+            wtype = 0; // 0 - "Huber", 1 - "BiSquare"
+
+            // Perform quadtree decomposition of frames
+            // to generate patches for noise estimation
+            int maxVsize = (Nx/size)*(Ny/size);
+            arma::vec means = -1 * arma::ones<arma::vec>(T*maxVsize);
+            arma::vec vars = -1 * arma::ones<arma::vec>(T*maxVsize);
+            arma::vec minslice(T);
+
+            // Perform quadtree decomposition of frames
+            // to generate patches for noise estimation
+            for (int i=0; i<T-1; i++) {                
```

Comment:
Error occurs here due to wrong size of `means` and `vars` - so currently just using **T-1** frames for noise estimation. Need to work out how to size correctly to fix.